src/Microsoft.AspNet.OData.Shared/Formatter/Serialization/ODataSerializerContext.cs

```diff
@@ -163,6 +165,35 @@ internal ODataQueryContext QueryContext
         /// </summary>
         public ODataPath Path { get; set; }
 
+        internal Type Type { get; set; }
+
+        internal bool IsUntyped
+        {
+            get
+            {
+                if (_isUntyped == null)
+                {
+                    _isUntyped = typeof(IEdmObject).IsAssignableFrom(Type) || typeof(EdmChangedObjectCollection).IsAssignableFrom(Type);
+                }
+
+                return _isUntyped.Value;
+            }
+        }
+
+        internal bool IsDeltaOfT
+        {
+            get
+            {
+                if (_isDeltaOfT == null)
+                {
+                    _isDeltaOfT = Type != null && TypeHelper.IsGenericType(Type) && (Type.GetGenericTypeDefinition() == typeof(DeltaSet<>) ||
```

Comment:
I don't know if `Type.GetGenericTypeDefinition()` is expensive but I'd probably start with the most commonly used when doing the type checks:
```csharp
(Type.GetGenericTypeDefinition() == typeof(Delta<>) || Type.GetGenericTypeDefinition() == typeof(DeltaSet<>) || Type.GetGenericTypeDefinition() == typeof(DeltaDeletedEntityObject<>))
```